src/main/java/org/spongepowered/common/world/schematic/SpongeSchematicBuilder.java

```diff
@@ -0,0 +1,158 @@
+import org.spongepowered.common.util.gen.ByteArrayMutableBlockBuffer;
+import org.spongepowered.common.util.gen.CharArrayMutableBlockBuffer;
+import org.spongepowered.common.util.gen.IntArrayMutableBlockBuffer;
+
+import java.util.Map;
+
+public class SpongeSchematicBuilder implements Schematic.Builder {
+
+    private ArchetypeVolume volume;
+    private Extent view;
+    private BlockPalette palette;
+    private BlockPaletteType type = BlockPaletteTypes.LOCAL;
+    private DataView metadata;
+    private Map<String, Object> metaValues = Maps.newHashMap();
+
+    @Override
+    public Builder volume(ArchetypeVolume volume) {
+        this.volume = volume;
+        return this;
+    }
+
+    @Override
+    public Builder volume(Extent volume) {
+        this.view = volume;
+        return null;
+    }
+
+    @Override
+    public Builder palette(BlockPalette palette) {
+        this.palette = palette;
+        this.type = palette.getType();
+        return this;
+    }
+
+    @Override
+    public Builder paletteType(BlockPaletteType type) {
+        this.type = type;
+        this.palette = null;
+        return this;
+    }
+
+    @Override
+    public Builder metadata(DataView metadata) {
+        this.metadata = metadata;
+        return this;
+    }
+
+    @Override
+    public Builder metaValue(String key, Object value) {
+        this.metaValues.put(key, value);
+        return this;
+    }
+
+    @Override
+    public Builder from(Schematic value) {
+        this.volume = value;
+        this.view = null;
+        this.palette = value.getPalette();
+        this.type = this.palette.getType();
+        this.metadata = value.getMetadata();
+        this.metaValues.clear();
+        return this;
+    }
+
+    @Override
+    public Builder reset() {
+        this.volume = null;
+        this.view = null;
+        this.palette = null;
+        this.type = BlockPaletteTypes.LOCAL;
+        this.metadata = null;
+        this.metaValues.clear();
+        return this;
+    }
+
+    @Override
+    public Schematic build() throws IllegalArgumentException {
+        if (this.palette == null) {
+            this.palette = this.type.create();
+        }
+        checkArgument(this.volume != null || this.view != null);
+        Vector3i min;
+        Vector3i size;
+        if (this.volume != null) {
+            min = this.volume.getBlockMin();
+            size = this.volume.getBlockSize();
+        } else {
+            min = this.view.getBlockMin();
+            size = this.view.getBlockSize();
+        }
+        Map<Vector3i, TileEntityArchetype> tiles = this.volume.getTileEntityArchetypes();
+        if (this.metadata == null) {
+            this.metadata = new MemoryDataContainer();
+        }
+        for (Map.Entry<String, Object> entry : this.metaValues.entrySet()) {
+            this.metadata.set(DataQuery.of(".", entry.getKey()), entry.getValue());
+        }
+        if (this.volume == null) {
```

Comment:
Another issue here... :tada: 
At line 141, there is a block handling the case of `this.volume` being null, but that's too late ! :laughing: 
Actually, `this.volume` is accessed before, at line 134, which throws a `NullPointerException`...
I think that actually line of code should be placed in a `else` statement, at line 154.